File: services/api/src/api/schemas/documents.py
from pydantic import BaseModel


class DocumentSaveResponse(BaseModel):
    document_id: int


# -------------------------
# POST: /api/documents
# -- Save raw text as document to database
# -------------------------
class DocumentTextSaveRequest(BaseModel):
    title: str
    src_lang: str
    text: str


# -------------------------
# GET: /api/documents
# -- Load document from database
# -------------------------
class DocumentImageOut(BaseModel):
    id: int
    media_type: str | None
    byte_length: int


class DocumentPageBlockOut(BaseModel):
    id: int
    block_index: int    
    block_type: str     # "text" | "image"
    tag: str | None
    text: str | None
    document_image_id: int | None
    alt: str | None


class DocumentSectionOut(BaseModel):
    id: int
    title: str  
    depth: int
    parent_section_id: int | None
    order_index: int
    first_page_number: int
    last_page_number: int


class DocumentPageOut(BaseModel):
    id: int
    page_number: int
    src_text: str
    section_id: int | None
    section_page_index: int | None
    blocks: list[DocumentPageBlockOut]


class DocumentLoadResponse(BaseModel):
    document_id: int
    title: str
    source_kind: str
    src_lang: str
    tgt_lang: str | None
    pages: list[DocumentPageOut]
    sections: list[DocumentSectionOut]
    images: list[DocumentImageOut]

File: apps/web/app/(protected)/documents/feature/api/documents.ts
import type { LoadedDocument, LoadedDocumentDTO } from "../types/document";


/**************************
 * `getDocumentById()`
 * -- GET: /api/documents
 *    Fetches Document from database given the document ID
 **************************/
export async function getDocumentById(
  documentId: number
): Promise<LoadedDocument> {
  // -------------------------
  // Fetch Document information
  // -------------------------
  const params = new URLSearchParams({ document_id: String(documentId) });
  const res = await fetch(`/api/documents?${params.toString()}`, {
    method: "GET",
    cache: "no-store",
  });

  if (!res.ok) {
    throw new Error("Failed to load document");
  }

  const data = (await res.json()) as LoadedDocumentDTO;

  // -------------------------
  // Sort pages
  // -- Preserve saved page blocks so the reader can interleave text + images
  // -------------------------
  const pages = (data.pages ?? []).map((p) => ({
    id: p.id,
    pageNumber: p.page_number,
    srcText: p.src_text,
    sectionId: p.section_id ?? null,
    sectionPageIndex: p.section_page_index ?? null,
    blocks: (p.blocks ?? []).map((block) => ({
      id: block.id,
      blockIndex: block.block_index,
      blockType: block.block_type,
      tag: block.tag ?? null,
      text: block.text ?? null,
      documentImageId: block.document_image_id ?? null,
      alt: block.alt ?? null,
    })),
  }));

  pages.sort((a, b) => a.pageNumber - b.pageNumber);

  // -------------------------
  // Preserve sections for current-page section title lookup
  // -------------------------
  const sections = (data.sections ?? []).map((s) => ({
    id: s.id,
    title: s.title,
    depth: s.depth,
    parentSectionId: s.parent_section_id,
    orderIndex: s.order_index,
    firstPageNumber: s.first_page_number,
    lastPageNumber: s.last_page_number,
  }));

  return {
    documentId: data.document_id,
    title: data.title,
    author: data.author,
    sourceKind: data.source_kind,
    coverImageId: data.cover_image_id ?? null,
    srcLang: data.src_lang,
    tgtLang: data.tgt_lang ?? null,
    pages,
    sections,
    // -------------------------
    // Preserve image metadata so page blocks can validate image references
    // -------------------------
    images: (data.images ?? []).map((image) => ({
      id: image.id,
      mediaType: image.media_type ?? null,
      byteLength: image.byte_length,
    })),
  };
}
